### PR TITLE
Reorganize repo and add signing/verifying interface

### PR DESCRIPTION
**Pull request by ltseeley**

### CODE REVIEW

File: libcylinder/src/lib.rs [hunk 1]
```diff
@@ -15,4 +16,118 @@
...
+
+impl PublicKey {
+    /// Creates a new public key
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self { bytes }
+    }
+
+    /// Returns the public key as a hex string
+    pub fn as_hex(&self) -> String {
+        hex::encode(&self.bytes)
```
Comment (agunde406) [inline libcylinder/src/lib.rs]: Can you use the local hex functions in this commit instead of adding it here and removing it later. (same for below)

File: libcylinder/src/signing/mod.rs [hunk 2]
```diff
@@ -212,6 +235,18 @@ impl<'a> Signer<'a> {
     }
 }
 
+impl<'a> super::Signer for Signer<'a> {
```
Comment (agunde406) [inline libcylinder/src/signing/mod.rs:241]: This seems like we should rename the existing signer to remove any confusion here.

File: libcylinder/src/lib.rs (hunk 1 above)
Comment (ltseeley) [reply libcylinder/src/lib.rs]: Updated

File: libcylinder/src/signing/mod.rs (hunk 2 above)
Comment (ltseeley) [reply libcylinder/src/signing/mod.rs]: Renamed to `ContextSigner`

File: libcylinder/src/signing/secp256k1.rs [hunk 3]
```diff
@@ -53,7 +52,9 @@ pub struct Secp256k1PrivateKey {
 
 impl Secp256k1PrivateKey {
     pub fn from_hex(s: &str) -> Result<Self, Error> {
-        hex_str_to_bytes(s).map(|key_bytes| Secp256k1PrivateKey { private: key_bytes })
+        hex_str_to_bytes(s)
+            .map(|key_bytes| Secp256k1PrivateKey { private: key_bytes })
+            .map_err(|err| Error::SigningError(Box::new(err)))
```
Comment (peterschwarz) [inline libcylinder/src/signing/secp256k1.rs]: This has changed the error type.  The previous error returned here was a `ParseError`, it also doesn't seem like you're using the enum created in the error module.

File: libcylinder/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (peterschwarz) [inline libcylinder/src/error.rs:1]: There's a lot of unrelated things happening in this commit. Error types created, functions moved that are unrelated to the primary purpose of the commit.

File: libcylinder/Cargo.toml [hunk 5]
```diff
@@ -39,6 +39,7 @@ experimental = [
   # The following features are experimental:
 ]
 
+hash = ["openssl"]
```
Comment (peterschwarz) [inline libcylinder/Cargo.toml:41]: Openssl seems like a pretty heavy dependency just for hashing.  I'd be open to converting this to using the [sha2](https://crates.io/crates/sha2) crate. (Can be left to a future PR).

File: libcylinder/src/signing/secp256k1.rs (hunk 3 above)
Comment (ltseeley) [reply libcylinder/src/signing/secp256k1.rs]: Good catch, switched it to `ParseError`. This method is removed in another commit.

File: libcylinder/src/error.rs (hunk 4 above)
Comment (ltseeley) [reply libcylinder/src/error.rs:1]: Split the commit in two: one for moving the hex functions and creating its error type, one for adding the signer/verifier interface and their error types.

File: libcylinder/Cargo.toml (hunk 5 above)
Comment (ltseeley) [reply libcylinder/Cargo.toml:41]: I'll make this change in a follow-up PR

**Comment by ltseeley** [comment]
> Given this removes the reading public and private keys from hex strings, is there a plan for adding that functionality back, or is that left to library callers?

I don't think we've specfically planned to add this functionality back, but we could pretty easily if it's desired.

**Comment by ltseeley** [comment]
Breaking up into smaller PRs
